kolibri/auth/models.py

```diff
@@ -26,120 +38,258 @@ class KolibriValidationError(KolibriError):
+
+    objects = SyncableModelPartitionQuerySet.as_manager()
+    # special reference to partition manager in case 'objects' is overridden in subclasses:
+    partitioned_objects = SyncableModelPartitionQuerySet.as_manager()
+
+    id = UUIDField(primary_key=True)
+
+    class Meta:
+        abstract = True
+
+    def calculate_uuid(self):
+        return uuid.uuid4().hex # for now, in this dummy stub, we're just using random UUIDs
+
+    def get_partition(self):
+        """
+        Calculate and return the partition attributes for the current model instance.
+        """
+        partition = {}
+        for partition_name, field_name in self._morango_partition_fields.items():
+            if hasattr(self, field_name):
+                partition[partition_name] = getattr(self, field_name)
+        return partition
+
+    def clean(self):
+        if not self.id:
+            self.id = self.calculate_uuid()
+        super(SyncableModel, self).clean()
+
+
+class MorangoTreeManager(TreeManager):
+
+    # Override the logic in Django MPTT that assigns tree ids, as we need to ensure tree ids do not
+    # conflict for facilities created on differentdevices and then later synced to the same device
+    # By default, tree ids are auto-increasing integers, but we use UUIDs to avoid collisions.
+    def _get_next_tree_id(self):
+        return uuid.uuid4().hex
+
+
+class MorangoMPTTModel(MPTTModel):
+    """
+    Any model that inherits from SyncableModel that wants to inherit from MPTTModel should instead inherit
+    from MorangoMPTTModel, which modifies some behavior to make it safe for the syncing system.
+    """
+
+    _default_manager = MorangoTreeManager()
+
+    # change tree_id to a uuid to avoid collisions; see explanation above in the MorangoTreeManager class
+    tree_id = UUIDField()
+
+    class Meta:
+        abstract = True
+
+
+##############################################
+# /END MORANGO
+##############################################
+
+
+class KolibriSyncableModel(SyncableModel):
+    """
+    Kolibri-specific subclass of Morango's SyncableModel class. Here, we add any additional fields needed
+    for partitioning the data (in this case, the "dataset" attribute), and specify the list of attributes
+    to be used for partitioning.
+    """
+
+    _morango_partition_fields = {
+        "dataset": "dataset_id",
+        "user": "user_id",
+    }
+
+    # in general, all syncable data must be associated with a FacilityDataset (but this can be overridden if needed)
+    _require_dataset = True
+
+    class Meta:
+        abstract = True
+
+    dataset = models.ForeignKey("FacilityDataset", blank=True, null=True)
+
+    def save(self, *args, **kwargs):
+        if not self.id:
+            self.id = self.calculate_uuid()
+        if self._require_dataset and not self.dataset:
+            raise IntegrityError("This model must be associated with a particular FacilityDataset.")
```

Comment:
Had a discussion with @benjaoming in the dev room about it being an anti-pattern to raise `ValidationError` in the `save` method. Since this case here is like a soft version of `null=False`, we figured a nice approach would be to raise a DB-level error like `IntegrityError`. A `ValidationError` is still raised in the `clean` method, so that it can be caught earlier and handled, if validation is taking place.